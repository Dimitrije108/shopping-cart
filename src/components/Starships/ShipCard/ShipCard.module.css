.linkWrapper {
  width: 100%;
}

.shipCard {
  width: 100%;
  min-width: 400px;
  height: auto;

  color: rgb(255, 255, 255);
  background-color: black;
	outline: 1px solid rgb(104, 104, 104);
	
  transition: all 0.3s ease-in-out;
  cursor: pointer;

  position: relative;
  display: flex;
  flex-direction: column;
}

.shipCard:hover,
.shipCard:active,
.shipCard:focus {
	outline: 1px solid rgb(221, 221, 221);
}
/* 3 arrow animation indicating additional information is available */
.showDetails {
  padding: 16px 12px;

  position: absolute;
  align-self: end;
  z-index: 1;
  visibility: hidden;

  opacity: 0;
  transition: opacity 0.5s ease-in-out;
}

.shipCard:hover > .showDetails,
.shipCard:active > .showDetails,
.shipCard:focus > .showDetails {
	visibility: visible;
  opacity: 1;
  transition-delay: 0s;
}

.arrow:nth-child(1) {
  transition-delay: 0.1s;
}

.arrow:nth-child(2) {
  transition-delay: 0.2s;
}

.arrow:nth-child(3) {
  transition-delay: 0.3s;
}

.imgCont {
  width: 100%;
  height: 100%;
  overflow: hidden;
  aspect-ratio: 16 / 9;

  display: flex;
  justify-content: center;
}

.imgCont > img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  object-position: center;
}

.infoCont {
  margin: 8px;
  padding: 4px 8px 8px 8px;
  outline: 1px solid rgb(104, 104, 104);

  display: grid;
  grid-template-areas: 
    "name price"
    "add add";
  gap: 1rem;
}

.nameCont {
  grid-area: name;
}

.name {
	margin: 0;

	font-size: 1.6rem;
  text-decoration: underline rgb(104, 104, 104);

  transition: all 0.2s ease-in-out;
}

.name:hover,
.name:active,
.name:focus {
  text-decoration: underline rgb(221, 221, 221);
}

.imperial {
  margin: 0;

  color: rgb(179, 179, 179);
  font-family: Imperial;
  font-size: 0.7rem;
  letter-spacing: 1.5px;
}

.naboo {
  margin: 0;

  color: rgb(179, 179, 179);
  font-family: Naboo;
  font-size: 0.5rem;
  letter-spacing: 1.5px;
}

.priceCont {
  grid-area: price;

  margin: 0;
  color: rgb(216, 215, 215);

  display: flex;
  flex-direction: column;
  align-items: flex-end;

  position: relative;
}

.priceCont > div {
  padding-bottom: 16px;
  font-family: "DINNext bold";
  font-size: 1.4rem;
  text-align: right;
}

.priceCont > p {
  margin: 0;
  padding: 0;

  font-family: DINNext;
  font-size: 1rem;
  text-align: right;

  position: absolute;
  top: 26px;
  right: 3px;
}

.addToCartCont {
  grid-area: add;
}

.addBtn {
  width: 100%;
  padding: 0.8rem 1rem;

  font-family: "DINNext bold";
  font-size: 1rem;
  color: rgb(204, 204, 204);
  border: 1px solid rgb(104, 104, 104);
}

.addBtn:hover,
.addBtn:active {
  color: rgb(255, 255, 255);
  border: 1px solid rgb(0, 192, 64);
}

.addBtn:focus{
  color: rgb(255, 255, 255);
}

.limitExceeded {
  width: 200px;
  padding: 6px;

  font-family: DINNext;
  font-size: 0.8rem;

  background-color: rgb(20, 20, 20);
  border: 1px solid rgb(104, 104, 104);
  cursor: auto;
  
  position: absolute;
  right: 180px;
  bottom: -45px;

  z-index: 1;
}

@media (max-width: 768px) {
  .shipCard {
    outline: 1px solid rgb(167, 167, 167);
  }
}

@media (max-width: 550px) {
  .shipCard {
    min-width: 310px;
  }

  .infoCont {
    grid-template-areas: 
      "name"
      "price"
      "add";
    gap: 0.5rem;
  }

  .priceCont {
    grid-area: price;
  
    display: flex;
    flex-direction: row;
    align-items: flex-end;
  
    position: static;
  }
  
  .priceCont > div {
    padding-bottom: 0;
    font-family: "DINNext bold";
    font-size: 1.2rem;
    text-align: left;
  }
  
  .priceCont > p {
    margin-left: 4px;
    padding-bottom: 1px;

    font-family: DINNext;
    font-size: 0.9rem;
    text-align: left;
  
    position: static;
  }
}

@media (hover: none) {
  .shipCard:hover {
    outline: none;
  }

  .name:hover {
    text-decoration: none;
  }

	.shipCard:hover > .showDetails {
    visibility: hidden;
  }
}
