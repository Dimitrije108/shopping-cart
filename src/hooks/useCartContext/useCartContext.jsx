import { useState, useEffect, createContext, useContext } from 'react';

const CartContext = createContext({
  cart: null,
	limitExceededMsg: "", 
  addToCart: () => {},
	removeFromCart: () => {},
});
// Shopping cart functionality passed on to all components 
// via context provider
export function useShoppingCart() {
	// Retrieve cart from storage if available
	const storedCart = JSON.parse(localStorage.getItem('cart'));
	const initCart = storedCart ? storedCart : [];

	const [cart, setCart] = useState(initCart);

	useEffect(() => {
		localStorage.setItem('cart', JSON.stringify(cart));
	}, [cart]);
	// Warning message for exceeding the ship quantity limit
	const limitExceededMsg = 
		"For orders larger than 20 ships per model please contact us directly!";
	// Add an item or adjust it's quantity
	const addToCart = (item) => {
		// check if the provided item is already inside the cart
		const isDuplicate = cart.find((ship) => ship.id === item.id);
		// handle duplicate items
		if (isDuplicate) {
			const totalQuantity = isDuplicate.quantity + item.quantity;
			// handle exceeding item quantity
			if (totalQuantity > 20) {
				return false;
			};
			// adjust quantity accordingly
			const addQuantity = cart.map((ship) => {
				if (ship.id === item.id) {
					return { ...ship, quantity: ship.quantity + item.quantity };
				}
				return ship;
			});
			setCart(addQuantity);
			return true;
		} else {
			// add unique item to the cart
			setCart([...cart, item]);
			return true;
		};
	};
	// Remove item instance from the cart
	const removeFromCart = (id) => {
		const removed = cart.filter((ship) => ship.id !== id);
		setCart(removed);
	};

	return { 
		cart, 
		limitExceededMsg, 
		addToCart, 
		removeFromCart,
	};
};

export const useCartContext = () => useContext(CartContext);

export default CartContext;