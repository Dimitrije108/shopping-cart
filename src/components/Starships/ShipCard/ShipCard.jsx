import { useState } from "react";
import { Link } from "react-router-dom";
import { useCartContext } from "../../../hooks/useCartContext/useCartContext";
import { useCartPopup } from "../../../hooks/useCartPopup/useCartPopup";
import formatNumber from "../../../formatNumber/formatNumber";
import HoverArrows from "../../HoverArrows/HoverArrows";
import styles from "./ShipCard.module.css";
// Create a ship card component
export default function ShipCard({ 
	id, 
	category, 
	name, 
	img, 
	price,
}) {
	const [tooltip, setTooltip] = useState(false);
	const { addToCart, limitExceededMsg } = useCartContext();
	const { addPopup } = useCartPopup();
	// Create link to view ship details
	const linkTo = `/starships/${category}/${id}`;
	// For instances where price returns "unknown"
	const truePrice = price === "unknown" 
		? "Price on Request" 
		: price;
	// Ship data for the shopping cart
	const cartItemInfo = {
		id,
		img,
		name,
		category,
		price: truePrice,
		quantity: 1,
	};
	// Handle adding the ship to the shopping cart
	const handleAdd = (e) => {
		e.preventDefault();
		e.stopPropagation();
		const added = addToCart(cartItemInfo);
		if (added) {
			addPopup(1, name);
		} else {
			setTooltip(true);
			setTimeout(() => {
				setTooltip(false);
			}, 3000);
		};
	};

	return (
		<Link to={linkTo}>
			<div className={styles.shipCard} data-testid="shipCard">
				<HoverArrows 
					showDetails={styles.showDetails}
					arrow={styles.arrow}
					width={15} 
					height={15}
				/>
				<div
					className={styles.imgCont}
					style={{ backgroundImage: `url(${img})` }}
					data-testid="imgCont"
				>
				</div>
				<div className={styles.infoCont}>
					<div className={styles.nameCont}>
						<h2 className={styles.name}>{name}</h2> 
						<div>
							<p className={styles.imperial}>{name}</p>
							<p className={styles.naboo}>{name}</p>
						</div>
					</div>
					<div className={styles.buyCont}>
						<div className={styles.priceCont}>
							<div>{formatNumber(truePrice)}</div>
							{truePrice !== "Price on Request" && <p>credits</p>}
						</div>
						<button 
							className={styles.addBtn}
							onClick={handleAdd}
						>
						ADD
						</button>
						{tooltip && 
							<p 
								className={styles.limitExceeded}
								onClick={(e) => {
									e.preventDefault();
									e.stopPropagation();
								}}
							>
									{limitExceededMsg}
							</p>
						}
					</div>
				</div>
			</div>
		</Link>
	)
};
