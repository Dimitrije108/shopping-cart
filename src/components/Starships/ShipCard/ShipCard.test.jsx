import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { useCartContext } from "../../../hooks/useCartContext/useCartContext";
import ShipCard from "./ShipCard";

vi.mock("../../../hooks/useCartContext/useCartContext", () => ({
	useCartContext: vi.fn(() => ({
		addToCart: vi.fn(),
	})),
}));

describe("Ship Card component", () => {
	it("ship card component snapshot", () => {
		const shipCard = render(
			<MemoryRouter>
				<ShipCard 
					id={'1234567890'}
					category={'starfighter'}
					name={'Millennium Falcon'}
					img={'https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg'}
					price={'100000'}
				/>
			</MemoryRouter>
		)

		expect(shipCard).toMatchSnapshot();
	});

	it("render elements inside the card", () => {
		render(
			<MemoryRouter>
				<ShipCard 
					id={'1234567890'}
					category={'starfighter'}
					name={'Millennium Falcon'}
					img={'https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg'}
					price={'100000'}
				/>
			</MemoryRouter>
		)

		const bckImg = screen.getByTestId('img-cont');
		const name = screen.getByRole('heading').textContent;
		const price = screen.getByText('100,000');
		const btn = screen.getByRole('button');

		expect(screen.getByRole('link')).toBeInTheDocument();
		expect(bckImg).toBeInTheDocument();
		expect(name).toMatch('Millennium Falcon');
		expect(price).toHaveTextContent('100,000');
		expect(btn).toBeInTheDocument();
	});

	it("redirect user to link destination", async () => {
		const user = userEvent.setup();

		render(
			<MemoryRouter initialEntries={['/starships']}>
				<Routes>
					<Route
						path="/starships"
						element={
							<ShipCard 
								id={'1234567890'}
								category={'starfighter'}
								name={'Millennium Falcon'}
								img={'https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg'}
								price={'100000'}
							/>
						}
					/>
					<Route path="/starships/:category/:id" element={<div>Starship Page</div>} />
				</Routes>
			</MemoryRouter>
		)

		const link = screen.getByRole('link');
		await user.click(link);

		expect(screen.getByText('Starship Page')).toBeInTheDocument();
	});

	it("add button adds item to cart", async () => {
		const user = userEvent.setup();

		const mockAddToCart = vi.fn();
		useCartContext.mockReturnValue({ addToCart: mockAddToCart});

		render(
			<MemoryRouter>
				<ShipCard 
					id={'1234567890'}
					category={'starfighter'}
					name={'Millennium Falcon'}
					img={'https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg'}
					price={'100000'}
				/>
			</MemoryRouter>
		)

		const addBtn = screen.getByRole('button', { name: "ADD" });
		await user.click(addBtn);

		expect(mockAddToCart).toHaveBeenCalledTimes(1);
	});

	it("render 'Price on Request' when price unknown", () => {
		render(
			<MemoryRouter>
				<ShipCard 
					id={'1234567890'}
					category={'starfighter'}
					name={'Millennium Falcon'}
					img={'https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg'}
					price={'unknown'}
				/>
			</MemoryRouter>
		)

		const price = screen.getByText('Price on Request');

		expect(price).toHaveTextContent('Price on Request');
	});
})