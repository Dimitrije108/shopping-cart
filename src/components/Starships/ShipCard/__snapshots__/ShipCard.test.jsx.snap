// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ship Card component > ship card component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        data-discover="true"
        href="/starships/starfighter/1234567890"
      >
        <div
          class="_shipCard_4857ee"
          data-testid="shipCard"
        >
          <div
            class="_showDetails_4857ee"
          >
            <img
              class="_arrow_4857ee"
              height="15"
              src="/src/assets/right-arrow.svg"
              width="15"
            />
            <img
              class="_arrow_4857ee"
              height="15"
              src="/src/assets/right-arrow.svg"
              width="15"
            />
            <img
              class="_arrow_4857ee"
              height="15"
              src="/src/assets/right-arrow.svg"
              width="15"
            />
          </div>
          <div
            class="_imgCont_4857ee"
            data-testid="imgCont"
            style="background-image: url(https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg);"
          />
          <div
            class="_infoCont_4857ee"
          >
            <div
              class="_nameCont_4857ee"
            >
              <h2
                class="_name_4857ee"
              >
                Millennium Falcon
              </h2>
              <div>
                <p
                  class="_imperial_4857ee"
                >
                  Millennium Falcon
                </p>
                <p
                  class="_naboo_4857ee"
                >
                  Millennium Falcon
                </p>
              </div>
            </div>
            <div
              class="_buyCont_4857ee"
            >
              <div
                class="_priceCont_4857ee"
              >
                <div>
                  100,000
                </div>
                <p>
                  credits
                </p>
              </div>
              <button
                class="_addBtn_4857ee"
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      data-discover="true"
      href="/starships/starfighter/1234567890"
    >
      <div
        class="_shipCard_4857ee"
        data-testid="shipCard"
      >
        <div
          class="_showDetails_4857ee"
        >
          <img
            class="_arrow_4857ee"
            height="15"
            src="/src/assets/right-arrow.svg"
            width="15"
          />
          <img
            class="_arrow_4857ee"
            height="15"
            src="/src/assets/right-arrow.svg"
            width="15"
          />
          <img
            class="_arrow_4857ee"
            height="15"
            src="/src/assets/right-arrow.svg"
            width="15"
          />
        </div>
        <div
          class="_imgCont_4857ee"
          data-testid="imgCont"
          style="background-image: url(https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg);"
        />
        <div
          class="_infoCont_4857ee"
        >
          <div
            class="_nameCont_4857ee"
          >
            <h2
              class="_name_4857ee"
            >
              Millennium Falcon
            </h2>
            <div>
              <p
                class="_imperial_4857ee"
              >
                Millennium Falcon
              </p>
              <p
                class="_naboo_4857ee"
              >
                Millennium Falcon
              </p>
            </div>
          </div>
          <div
            class="_buyCont_4857ee"
          >
            <div
              class="_priceCont_4857ee"
            >
              <div>
                100,000
              </div>
              <p>
                credits
              </p>
            </div>
            <button
              class="_addBtn_4857ee"
            >
              ADD
            </button>
          </div>
        </div>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
